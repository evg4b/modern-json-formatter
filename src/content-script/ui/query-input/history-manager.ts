export class HistoryManager<T> {
  private readonly history: T[] = [];
  private historyIndex = -1;

  save(state: T) {
    if (this.historyIndex < this.history.length - 1) {
      this.history.splice(this.historyIndex + 1);
    }

    this.history.push(state);
    this.historyIndex = this.history.length - 1;
  }

  undo() {
    if (this.historyIndex > 0) {
      this.historyIndex--;
      return this.history[this.historyIndex];
    }
    return null;
  }

  redo() {
    if (this.historyIndex < this.history.length - 1) {
      this.historyIndex++;
      return this.history[this.historyIndex];
    }
    return null;
  }
}
