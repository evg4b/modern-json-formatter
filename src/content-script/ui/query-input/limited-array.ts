export class LimitedArray<T> extends Array<T> {
  constructor(private readonly limit: number) {
    super();
  }

  public push(...items: T[]): number {
    if (this.length >= this.limit) {
      this.shift();
    }
    return super.push(...items);
  }
}
