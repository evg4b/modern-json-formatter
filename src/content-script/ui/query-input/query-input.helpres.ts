export const isUndoEvent = (event: KeyboardEvent) => {
  return (event.metaKey || event.ctrlKey) && event.key === 'z' && !event.shiftKey;
};

export const isRedoEvent = (event: KeyboardEvent) => {
  return (event.metaKey || event.ctrlKey) && event.key === 'z' && event.shiftKey;
};

export const isWrapEvent = (event: KeyboardEvent, brackets: Record<string, string>) => {
  return event.key in brackets;
};

export const isSubmitEvent = (event: KeyboardEvent) => {
  return event.key === 'Enter';
};

export const isPrintableKey = (event: KeyboardEvent) => {
  return event.key.length === 1
    && !event.shiftKey
    && !event.ctrlKey
    && !event.metaKey
    && !event.altKey;
}
