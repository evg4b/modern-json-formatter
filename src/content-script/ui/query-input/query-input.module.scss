:host {
  display: flex;
  flex-direction: row;
  gap: 5px;

  input {
    min-width: 200px;
    width: 30vw;
  }

  .input-wrapper {
    display: flex;
    flex-direction: column;
    gap: 3px;
    position: relative;

    .error-message {
      position: absolute;
      top: calc(100% + 5px);
      color: red;
      font-size: 10px;
      user-select: none;

      background: var(--background-color);
      padding: 2px;
      box-shadow: 0 0 2px 1px var(--background-color);

      &.hidden {
        display: none;
      }
    }
  }
}
