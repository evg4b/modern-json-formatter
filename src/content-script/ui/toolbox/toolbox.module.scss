:host {
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  position: fixed;
  top: 15px;
  right: 30px;
}
