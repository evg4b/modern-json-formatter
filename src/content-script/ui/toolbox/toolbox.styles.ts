export default `
:host {  
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  position: fixed;
  top: 0;
  right: 0;
  padding: 10px 20px 5px 5px;
  background: var(--background);
  box-shadow: 0 0 3px 3px var(--background);
  gap: 10px;

  .button-container {
    display: flex;
    flex-direction: row;
    gap: 5px;

    button {      
      border-radius: 5px;
      padding: 3px 10px;
      background: var(--button-background);
      color: var(--button-color);
      border: 1px solid var(--button-border-color);
      cursor: pointer;
      font-size: 14px;
      font-weight: 500;
      min-width: 40px;
      display: block;
      user-select: none;
      transition-property: background, background-color, border-color, color;
      transition-duration: 0.2s;
      transition-timing-function: ease-in-out;

      &:hover {
        background: var(--button-hover-background);
        border-color: var(--button-hover-border-color);
        color: var(--button-hover-color);
      }

      &.active {
        background: var(--button-active-background);
        border-color: var(--button-active-border-color);
        color: var(--button-active-color);
        cursor: default;
      }
    }
  }
}
`
