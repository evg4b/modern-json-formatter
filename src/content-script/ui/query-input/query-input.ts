import { getHistory, pushHistory } from '@core/background';
import { getURL } from '@core/browser';
import { createElement, CustomElement, StyledComponentElement } from '@core/dom';
import { debounce } from 'lodash';
import { isNotNil } from '../../helpers';
import { InfoButtonElement } from '../info-button';
import { QueryHistory } from '../query-history/query-history';
import { HistoryManager } from './history-manager';
import { isRedoEvent, isSubmitEvent, isUndoEvent, isWrapEvent } from './query-input.helpers';
import queryInputStyles from './query-input.module.scss';

export const brackets: Record<string, string> = {
  '[': ']',
  '(': ')',
  '{': '}',
  '\'': '\'',
  '"': '"',
  '`': '`',
};

interface HistoryItem {
  query: string;
  start: number | null;
  end: number | null;
}

@CustomElement('query-input')
export class QueryInputElement extends StyledComponentElement {
  private readonly inputElement = this.createInput();
  private readonly errorMessageElement = createElement({
    element: 'span',
    class: ['error-message', 'hidden'],
  });
  private readonly queryHistoryElement = new QueryHistory();
  private readonly wrapperElement = createElement({
    element: 'div',
    class: 'input-wrapper',
    children: [
      this.inputElement,
      this.errorMessageElement,
      this.queryHistoryElement,
    ],
  });

  private readonly infoIcons = new InfoButtonElement(getURL('faq.html'));
  private readonly history = new HistoryManager<HistoryItem>();

  private onSubmitCallback: ((s: string) => unknown) | null = null;

  constructor() {
    super(queryInputStyles);
    this.shadow.append(this.infoIcons, this.wrapperElement);
    this.setupEventHandlers(this.inputElement);
    this.saveState();
    this.queryHistoryElement.onSelectedQuery(query => {
      this.inputElement.value = query;
      this.onSubmitEvent();
    });
  }

  public setErrorMessage(errorMessage: string | null): void {
    if (errorMessage) {
      this.errorMessageElement.textContent = errorMessage;
      this.errorMessageElement.classList.remove('hidden');
    } else {
      this.errorMessageElement.textContent = '';
      this.errorMessageElement.classList.add('hidden');
    }
  }

  public onSubmit(callback: (s: string) => void | Promise<void>): void {
    this.onSubmitCallback = callback;
  }

  public focus(): void {
    this.inputElement.focus();
  }

  public blur(): void {
    this.inputElement.blur();
  }

  public hide(): void {
    this.style.display = 'none';
  }

  public show(): void {
    this.style.display = 'flex';
  }

  private createInput(): HTMLInputElement {
    const input = createElement({ element: 'input' });
    input.type = 'text';
    input.placeholder = 'Input jq query...';

    return input;
  }

  private setupEventHandlers(input: HTMLInputElement) {
    input.addEventListener('keydown', event => {
      this.setErrorMessage(null);
      if (isSubmitEvent(event)) {
        this.onSubmitEvent();
      } else if (isWrapEvent(event, brackets)) {
        this.onWrapEvent(event);
      } else if (isUndoEvent(event)) {
        this.onUndoEvent(event);
      } else if (isRedoEvent(event)) {
        this.onRedoEvent(event);
      }
    });

    input.addEventListener('input', () => {
      this.saveState();
      void this.loadHistory(this.inputElement.value);
      console.log('input', this.inputElement.value);
    });

    input.addEventListener('focus', () => {
      this.queryHistoryElement.open();
    });
  }

  private onSubmitEvent() {
    this.onSubmitCallback?.(this.inputElement.value);
    this.queryHistoryElement.close();
    pushHistory(window.location.hostname, this.inputElement.value);
  }

  private readonly loadHistory = debounce(async (prefix: string) => {
    const history = await getHistory(window.location.hostname, prefix);
    this.queryHistoryElement.setHistory(history);
    this.queryHistoryElement.open();
  }, 300);

  private onWrapEvent(event: KeyboardEvent) {
    const start = this.inputElement.selectionStart ?? 0;
    const end = this.inputElement.selectionEnd ?? 0;
    if (isNotNil(start) && isNotNil(end) && start !== end) {
      event.preventDefault();
      const selectedText = this.inputElement.value.substring(start, end);
      this.inputElement.setRangeText(`${ event.key }${ selectedText }${ brackets[event.key] }`);
      this.inputElement.setSelectionRange(start + 1, end + 1);
      this.saveState();
    }
  }

  private onUndoEvent(event: KeyboardEvent) {
    event.preventDefault();
    const previousState = this.history.undo();
    this.restoreState(previousState);
  }

  private onRedoEvent(event: KeyboardEvent) {
    event.preventDefault();
    const nextState = this.history.redo();
    this.restoreState(nextState);
  }

  private saveState() {
    this.history.save({
      query: this.inputElement.value,
      start: this.inputElement.selectionStart,
      end: this.inputElement.selectionEnd,
    });
  }

  private restoreState(state: HistoryItem | null) {
    if (state) {
      this.inputElement.value = state.query;
      this.inputElement.setSelectionRange(state.start, state.end);
    }
  }
}
