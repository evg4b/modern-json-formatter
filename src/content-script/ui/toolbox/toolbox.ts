import { createElement, CustomElement, StyledComponentElement } from '@core/dom';
import { assetTabType } from '../../helpers';
import { QueryInputElement } from '../query-input';
import toolboxStyles from './toolbox.styles.txt';

@CustomElement('extension-toolbox')
export class ToolboxElement extends StyledComponentElement {
  private readonly input = new QueryInputElement();
  private readonly rawButton = this.createButton('Raw', 'raw');
  private readonly formattedButton = this.createButton('Formatted', 'formatted', true);
  private readonly queryButton = this.createButton('Query', 'query');

  private readonly buttonList: [TabType, HTMLButtonElement][] = [
    ['raw', this.rawButton],
    ['formatted', this.formattedButton],
    ['query', this.queryButton],
  ];

  private tabChangedCallback: ((s: TabType) => void) | null = null;

  constructor() {
    super(toolboxStyles);
    const container = createElement({
      element: 'div',
      class: 'button-container',
      children: [this.queryButton, this.formattedButton, this.rawButton],
    });
    this.shadow.append(this.input, container);
    this.shadow.addEventListener('click', e => {
      if (e.target instanceof HTMLButtonElement) {
        e.preventDefault();
        const ref = e.target.getAttribute('ref');
        assetTabType(ref);
        this.activateButton(ref);
      }
    });
    this.input.hide();
  }

  public onQueryChanged(callback: (s: string) => void | Promise<void>): void {
    this.input.onSubmit(callback);
  }

  public onTabChanged(callback: (s: TabType) => void): void {
    this.tabChangedCallback = callback;
  }

  public setErrorMessage(error: string): void {
    this.input.setErrorMessage(error);
  }

  private activateButton(tab: TabType): void {
    this.buttonList.forEach(([key, value]) =>
      key === tab ? value.classList.add('active') : value.classList.remove('active'),
    );
    this.tabChangedCallback?.(tab);
    if (tab === 'query') {
      this.input.show();
      this.input.focus();
    } else {
      this.input.hide();
    }
  }

  private createButton(content: string, ref: TabType, active = false): HTMLButtonElement {
    return createElement({
      element: 'button',
      content,
      class: active ? 'active' : undefined,
      attributes: { type: 'button', ref },
    });
  }
}
