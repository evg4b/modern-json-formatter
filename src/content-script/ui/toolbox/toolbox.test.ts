/* eslint-disable @typescript-eslint/unbound-method */
import '@testing/browser.mock';
import '@testing/query-input.mock';
import { beforeEach, describe, expect, test } from '@jest/globals';
import { getShadowRoot } from '@testing/styled-component';
import { type QueryInputElement } from '../query-input';
import { ToolboxElement } from './toolbox';

describe('Toolbox', () => {
  let toolbox: ToolboxElement;
  let shadowRoot: ShadowRoot;

  beforeEach(() => {
    toolbox = new ToolboxElement();
    shadowRoot = getShadowRoot(toolbox);
  });

  describe('button', () => {
    const cases = [
      { name: 'raw', selector: `button[ref="raw"]` },
      { name: 'formatted', selector: `button[ref="formatted"]` },
      { name: 'query', selector: `button[ref="query"]` },
    ];

    describe.each(cases)(`$name`, ({ name, selector }) => {
      let button: HTMLButtonElement | null;

      beforeEach(() => {
        button = shadowRoot.querySelector<HTMLButtonElement>(selector);
      });

      test('should exists', () => {
        expect(button).toBeTruthy();
      });

      test(`should match to snapshot`, () => {
        expect(button).toMatchSnapshot();
      });

      test(`should have handle click`, () => {
        const handler = jest.fn();
        toolbox.onTabChanged(handler);
        button?.click();

        expect(handler).toHaveBeenCalledWith(name);
      });
    });
  });

  describe('input', () => {
    let input: QueryInputElement | null;

    beforeEach(() => {
      input = shadowRoot.querySelector<QueryInputElement>('query-input');
    });

    test('should exists', () => {
      expect(input).toBeTruthy();
    });

    test('should be hided by default', () => {
      expect(input?.hide).toHaveBeenCalled();
    });

    describe('after query button click', () => {
      beforeEach(() => {
        shadowRoot.querySelector<HTMLButtonElement>('button[ref="query"]')?.click();
      });

      test('should be visible', () => {
        expect(input?.show).toHaveBeenCalledTimes(1);
      });

      test('should be focused', () => {
        expect(input?.focus).toHaveBeenCalledTimes(1);
      });

      test('should be hidden after raw button click', () => {
        shadowRoot.querySelector<HTMLButtonElement>('button[ref="raw"]')?.click();
        expect(input?.hide).toHaveBeenCalledTimes(2);
      });
    });

    test(`should set error to the input`, () => {
      const errorMessage = 'Test error message';

      toolbox.setErrorMessage(errorMessage);

      expect(input?.setErrorMessage).toHaveBeenCalledWith(errorMessage);
    });

    test(`should set submit handler on the input`, () => {
      const handler = () => {
        // stub for test
      };

      toolbox.onQueryChanged(handler);

      expect(input?.onSubmit).toHaveBeenCalledWith(handler);
    });
  });
});
