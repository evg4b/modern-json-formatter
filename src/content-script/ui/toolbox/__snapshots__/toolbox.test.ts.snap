// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbox button formatted should match to snapshot 1`] = `
<button
  class="active"
  ref="formatted"
  type="button"
>
  Formatted
</button>
`;

exports[`Toolbox button query should match to snapshot 1`] = `
<button
  ref="query"
  type="button"
>
  Query
</button>
`;

exports[`Toolbox button raw should match to snapshot 1`] = `
<button
  ref="raw"
  type="button"
>
  Raw
</button>
`;
